### PR TITLE
Webvr to webxr

### PR DESCRIPTION
**Pull request by catmajor**
Converts webVR API to webXR for fish tank

Demo at https://catmajor.github.io/WebGLSamples.github.io/aquarium/aquarium.html

### CODE REVIEW

**Comment by kainino0x** [comment]
@toji?

**Review by toji — CHANGES_REQUESTED** [review]
Thank you for taking on this update! I've wanted to do it myself for a while and never had the time. It's a big undertaking, and there's some subtle and easy-to-trip-over quirks to working in VR in any environment, so please don't take this review as a knock on your code. I'm just trying to focus on the possible issues so that this can be a comfortable, fun experience for those who try it!

I tried this both on an Android phone with Cardboard and on Quest 3, and it's failing differently on both devices.

On Cardboard things initially look OK but there's no stereo depth to the scene? It feels like everything is projected on a dome very far away. Also, looking a little to the left and right feels OK but when turning around 360 degrees the view stops responding, locks up, and eventually clicks back into place again. It indicates something is being done wrong in the view matrix math.

On the Quest 3 I initially only got a blue background with some bubbles rendering when I first tried to enter VR. The second time I saw the fish but the eye offsets were wrong and I couldn't converge the two images, which was quite uncomfortable. The third time I tried I was back to the blue backgrounds with a random bubble effect as the only thing visible.

When doing development I highly recommend testing the full range of head motion on at least an Android device in cardboard mode. Even if you don't have a cardboard viewer available you'll be able to observe areas where the camera motion doesn't match the rotation of the device.

Also, if you don't have a device like a Quest handy I would advise not attempting to include features that only work in their browser, like multiview. That can come as a follow-up if needed.

I've added some comments to specific areas of the code that I think are likely to be misbehaving at the moment. Feel free to ask further questions or toss alternate ideas my way and we can discuss them as you go! Again, this is a complex API with lots of unfortunate sharp edges, so don't get discouraged that things aren't working as expected on the first try.

**Comment by catmajor** [comment]
@toji 
Thank you Brandon for your review. I will address those issues over the weekend.

File: aquarium/aquarium.js
Comment (catmajor) [reply aquarium/aquarium.js]: Tried it on Oculus Browser and Wolvic on Quest 2, and it does not work. Going to investigate it.
Tested on cardboard on Galaxy A34 on chrome versions 120.0.6099.145 and .193 .
Previously, I had problems with older versions of Chrome crashing on entering XR, but this does not happen anymore.

I made this change but later removed this code because it was unnecessary since all matrices were coming from xrPose.

File: aquarium/aquarium.js
Comment (catmajor) [reply aquarium/aquarium.js]: Made this change but removed later because it was unnecessary since all matrices were coming from xrPose.

File: aquarium/aquarium.js
Comment (catmajor) [reply aquarium/aquarium.js]: ditto

File: aquarium/aquarium.js
Comment (catmajor) [reply aquarium/aquarium.js]: Done

Also removed the obsolete calculateViewMatrix block.

**Comment by toji** [comment]
Has the link in the first comment been updated with the latest changes? It seems different at least.

Assuming it has been, I now see with a Cardboard viewer that the view orientation is inverted. (As in, if you look up the virtual view looks down, if you look left the virtual view looks right.)

Looking into the code, I can see in the `render()` function implementation that the `viewInverse` variable is, ironically enough, being inverted and then multiplied with the projection matrix to make the final projection+view matrix used by the shaders. Since the "view" matrix in most rendering is, effectively, the inverse of the camera transform, this feels like the naming is backwards to me. Regardless, it suggests that you should maybe be using `view.transform.matrix` instead of `view.transform.inverse.matrix`.

Sorry for the possibly misleading initial suggestion, I blame the unclear variable names. 🤷 

https://github.com/WebGLSamples/WebGLSamples.github.io/blob/f2a9a34d46acb57677d8a1d87e3f5e6e24f2c986/aquarium/aquarium.js#L1290

**Comment by catmajor** [comment]
Hi Brandon, 
Sorry for the delay with the reply. I was puzzled as to why it wasn't working on Oculus. 
However, after a recent Oculus update, it looks like it works in Oculus browser version 122.0.6261.64 .
Wolvic browser still doesn't work.
Thanks,
Nikita

File: aquarium/aquarium.js
Comment (toji) [inline aquarium/aquarium.js:1782]: Outdated comment?

File: aquarium/aquarium.js
Comment (toji) [inline aquarium/aquarium.js:1789]: Let's removed these commented out calls unless they're trying to demonstrate some alternate technique. If so, they should be commented to indicate what they are trying to show.

File: aquarium/aquarium.js
Comment (toji) [inline aquarium/aquarium.js:2188]: I suspect that this will eventually want to be changed from `'local'` to `'local-floor'` to get some of the effect I was talking about with aligning to the user's floor. It probably won't make much visual difference at the moment, though, given the scale issues.

**Review by toji — COMMENTED** [review]
This is looking much better! Tested with Cardboard and a Pixel 8 just now and the head tracking is working as expected. Thanks for pushing it this far! I think that there's some additional work to do in terms of presenting the content well, but that can probably come as a separate PR.

Specifically, while the head tracking is working correctly it looks like the fish tank itself is gigantic relative to the viewer. The viewport is positioned very low to the floor and the fish seem very large and far away, meaning that any stereo effect is minimal. I haven't tested with my Quest headsets yet, but I suspect that walking around would yield very little movement relative to the tank floor because of the apparent scale involved. I suspect that if the tank was scaled down or the viewer's movements scaled up (I'm not yet sure which would be easier) then the effect would be much more striking. Ideally we probably want the user's head to end up somewhere near the center of the aquarium by default, with the virtual floor lining up with their physical one. That would likely still make the fish somewhat large relative to the user, but they'd be at a scale and distance that makes them much easier to observe and feel like you're in the same space with. But like I said, that can come later. The fact that this works at all means it's a big improvement over the old, deprecated WebVR code.

I will say that I still don't quite understand the motive behind the "Stereo mode". The two viewports it presents appear to be identical, so it doesn't provide a stereo effect even if you cross your eyes or have some other way of combining the images. And when you do the page UX still overlaps only one side of the canvas, which would interfere with any stereo effect that was present. I'd prefer to remove that option unless there's a known environment in which it's beneficial (custom kiosk, maybe?) and if that's the case it can probably be only shown with a specific URL argument.

I'm not sure what would be wrong with Wolvic, but I suspect their implementation is still considered a WIP. I'll see if I can test it later and dig up what the problematic part is.

In any case, lets either remove or situationally hide the Stereo View button and address a couple of other comments and I think we can land this!

File: aquarium/aquarium.js
Comment (catmajor) [reply aquarium/aquarium.js]: Also fixed

**Comment by catmajor** [comment]
Hi Brandon, 
I adjusted the position scale to be closer to the real world and eye position to be on the fishs' level. 
Also scaled the fishes down a bit. 
I changed the stereo mode button to only appear when stereo-enabled=true in the URL.

I have a question about VR UI. It resides now in aquarium-vr.
I'd like to move it to aquarium folder and try to make it work in another PR.
What do you think?

Thanks,
Nikita

**Review by toji — APPROVED** [review]
This looks great now! The scale changes you made do help quite a bit, as the fish and the elements in the tank clearly feel like they have depth now. Testing with a Quest 3 I can move around and the tank and fish have a nice scale and position to them.

I know this was a pain to work through, but I really appreciate you sticking with it and working out the bugs. It's really cool to be able to experience this classic WebGL demo in a new way! Thank you!

**Comment by toji** [comment]
I'm not entirely sure what you mean by the VR UI in aquarium-vr? I've never tried it. I'd generally say that unless there's a large amount of code differences between the two, however, it seems like a reasonable idea to pull any remaining changes over into this one and have just the single "aquarium" folder.

Also, every WebXR-capable device can support a `local-floor` reference space, but you have to request an additional feature that I forgot to mention. In any case, I tested it out locally and it does help with the user's position quite a bit! (Puts your head approximately in the center of the tank, which is nice.) The floor doesn't naturally line up, though, which feels a little disconcerting. I think once this lands I'll take a pass at adjusting the reference space and getting the floor to line up so that people with 6DoF headsets feel more grounded. Regardless, it's a fun demo even without those adjustments so I think this should be landed as-is!

**Comment by catmajor** [comment]
Thanks Brandon,
I'll play more with local floor. 
There's some sort of VR UI in the aquarium-vr folder (vr_assets/ui). It looks like VR code was taken from aquarium-vr. It refers to VR UI but fails because the folder does not exist. I'd like to move it and make it functional.

**Comment by toji** [comment]
Got it. I'd be interested to see what that entails. In the meantime I still think this can land as-is with the other tasks you're talking about being left to follow-up PRs.

@kainino0x, do you have write access to this repo? If not, who should we ping about merging?

**Comment by kenrussell** [comment]
Thanks @toji for reviewing. I invited you to be a co-owner of this repo, but merging this myself. Thanks @catmajor for your contribution.

**Comment by GKraats** [comment]
<pre>Since a short time aquarium totally fails if webgl2 not available.
I am running an old Dell laptop with linux Debian Testing 
and firefox, which does not support webgl2.
It only shows a page with misleading errormessage:
&quot;It does not appear your computer supports WebGL&quot;
and loops with varying fps-count.

Problem is caused by this Merge-Request.
At aquarium/aquarium-config.js it sets enableVR: true, 
causing webgl2 to become the preferred driver.
The first check for driver webgl2 fails, causing the 
display of the error-page.
The next check for driver webgl succeeds, causing the 
program to continue, but probably because of the active error-page,
it fails displaying the aquarium.

Problem can be solved at tdl/webgl.js at
tdl.webgl.setupWebGL.
Function handleCreationError should only
generate the message but not show it.
The message should be only shown at the end of
tdl.webgl.setupWebGL, if no context is available.
</pre>

**Pull request merged**
